lambda_metadata.tf

```diff
@@ -0,0 +1,181 @@
+  policy_arn = aws_iam_policy.metadata_removal_lambda[count.index].arn
+  roles      = [aws_iam_role.metadata_removal_lambda[count.index].name]
+}
+
+data "aws_iam_policy_document" "metadata_removal_lambda" {
+  count = local.enable_metadata_purge_lambda[local.environment] ? 1 : 0
+  statement {
+    sid = "LogGroup"
+    actions = [
+      "logs:CreateLogGroup",
+      "logs:CreateLogStream",
+      "logs:PutLogEvents"
+    ]
+    resources = [
+      "*"
+    ]
+  }
+
+  statement {
+    sid = "EMRDescribeClusters"
+    actions = [
+      "elasticmapreduce:DescribeCluster"
+    ]
+    resources = ["*"]
+  }
+
+  statement {
+    sid = "HBaseRootDirOperations"
+    actions = [
+      "s3:GetObject",
+      "s3:DeleteObject",
+    ]
+    resources = [
+      "arn:aws:s3:::${local.hbase_root_bucket}/${data.terraform_remote_state.ingest.outputs.ingest_emr_s3_prefixes["base_root_prefix"]}/*",
+      "arn:aws:s3:::${local.hbase_root_bucket}/${data.terraform_remote_state.ingest.outputs.ingest_emr_s3_prefixes["base_root_prefix"]}/",
+    ]
+  }
+
+  statement {
+    sid = "HBaseBucketOperations"
+    actions = [
+      "s3:ListBucket*",
+    ]
+    resources = [
+      "arn:aws:s3:::${local.hbase_root_bucket}",
+    ]
+  }
+
+}
+
+
+resource "aws_iam_policy" "metadata_removal_lambda" {
+  count       = local.enable_metadata_purge_lambda[local.environment] ? 1 : 0
+  name        = "metadata-removal-lambda"
+  description = "policy for lambda function that deletes replica metadata from hbase dir"
+  policy      = data.aws_iam_policy_document.metadata_removal_lambda[count.index].json
+
+  tags = { Name = "metadata-removal-lambda-policy" }
+}
+
+data "archive_file" "metadata_removal_lambda" {
+  count       = local.enable_metadata_purge_lambda[local.environment] ? 1 : 0
+  type        = "zip"
+  source_dir  = "files/metadata_removal_lambda"
+  output_path = "files/metadata_removal_lambda.zip"
+}
+
+resource "aws_lambda_function" "metadata_removal" {
+  count            = local.enable_metadata_purge_lambda[local.environment] ? 1 : 0
+  filename         = data.archive_file.metadata_removal_lambda[count.index].output_path
+  source_code_hash = data.archive_file.metadata_removal_lambda[count.index].output_base64sha256
+  function_name    = "metadata_removal_lambda"
+  role             = aws_iam_role.metadata_removal_lambda[count.index].arn
+  description      = "Lambda function to remove replica metadata from hbase directory"
+  handler          = "index.handler"
+  runtime          = "python3.8"
+  timeout          = 900
+
+  tags = { Name = "metadata-removal-lambda" }
+}
+
+resource "aws_cloudwatch_event_rule" "emr_state_change" {
+  count       = local.enable_metadata_purge_lambda[local.environment] ? 1 : 0
+  name        = "emr-status-change"
+  description = "captures status changes of emr clusters"
+
+  event_pattern = jsonencode(
+    {
+      "source" : [
+        "aws.emr"
+      ],
+      "detail-type" : [
+        "EMR Cluster State Change"
```

Comment:
This feels like you'll be triggering unnecessary lambda executions with doing the checks for cluster name and status in the lambda instead. You could leave that code in for extra security if you want, but I'd still make the cloudwatch event more specific by adding the name here.

Could have two event rules, one for TERMINATED and one for TERMINATED_WITH_ERRORS. This is what a lot of our clusters do for their alerts -> https://github.com/dwp/aws-analytical-dataset-generation/blob/master/slack-alerts.tf#L1.

Reply:
Agreed, I did mean to resolve this